File: src/engine/supportBundle.ts
import type {
  FeedHealth,
  OutputDestinationState,
  Participant,
  ProductionState,
  SceneTemplate,
  SupportBundle,
  SupportBundleMediaCore
} from "../domain/production";
import type { SupportBundleContext } from "./contracts";
import type { NativeMediaCoreStateSnapshot } from "./nativeMediaCoreProtocol";
import { buildRecordingManifest } from "./recordingManifest";

const MOCK_FREE_DISK_BYTES = 86 * 1024 * 1024 * 1024;

export class InMemoryDiagnosticsEngine {
  async createSupportBundle(
    state: ProductionState,
    mediaCore?: NativeMediaCoreStateSnapshot,
    context?: SupportBundleContext
  ): Promise<SupportBundle> {
    return createSupportBundle(state, mediaCore, context);
  }
}

export function createSupportBundle(
  state: ProductionState,
  mediaCore?: NativeMediaCoreStateSnapshot,
  context: SupportBundleContext = {}
): SupportBundle {
  const createdAt = new Date().toISOString();
  const duplicateAssignments = countDuplicateAssignments(state.scenes);
  const unavailableScreenShare = countUnavailableScreenShare(state);
  const lowDeliveredResolution = state.participants.filter((participant) => participant.health === "low-resolution").length;
  const warnings = buildWarnings(state, duplicateAssignments, unavailableScreenShare, lowDeliveredResolution);
  const freeDiskBytes = context.freeDiskBytes ?? MOCK_FREE_DISK_BYTES;
  const isoCapacity = estimateIsoCapacity(state, freeDiskBytes);
  const destinations = getDestinationStates(state);
  const triageLines = buildTriageLines(state, warnings, isoCapacity, context);

  return {
    id: `support-${slugify(state.meetingTitle)}-${createdAt.replace(/[-:.TZ]/g, "").slice(0, 14)}`,
    createdAt,
    app: {
      name: "CoreVideo Pro",
      version: context.version ?? "0.1.0",
      platform: context.platform ?? "mock-desktop"
    },
    summaryText: triageLines.join("\n"),
    triageLines,
    production: {
      meetingTitle: state.meetingTitle,
      mode: state.mode,
      activeSceneId: state.activeSceneId,
      previewSceneId: state.previewSceneId,
      selectedBreakoutRoomId: state.selectedBreakoutRoomId,
      recording: state.recording,
      streaming: state.streaming
    },
    participants: state.participants.map((participant) => ({
      id: participant.id,
      name: participant.name,
      role: participant.role,
      health: participant.health,
      breakoutRoomName: participant.breakoutRoomName,
      isActiveSpeaker: participant.isActiveSpeaker,
      isScreenSharing: participant.isScreenSharing,
      muted: participant.isMuted,
      recommendedAction: getParticipantAction(participant)
    })),
    output: {
      health: { ...state.outputSession.health },
      activeDestinationIds: [...state.outputSession.activeDestinationIds],
      destinations: destinations.map((destination) => ({
        id: destination.id,
        name: destination.name,
        protocol: destination.protocol,
        enabled: destination.enabled,
        active: destination.active,
        health: destination.health,
        bitrateMbps: destination.bitrateMbps,
        endpoint: redactEndpoint(destination.endpoint),
        streamKey: redactSecret(destination.streamKey)
      })),
      statusText: state.outputSession.statusText
    },
    actionCounts: {
      duplicateAssignments,
      unavailableScreenShare,
      lowDeliveredResolution,
      sdkSubscribeErrors: 0
    },
    isoCapacity,
    mediaCore: mediaCore ? summarizeMediaCore(mediaCore, state) : undefined,
    runtime: context.runtime,
    crashRecovery: context.crashEvents?.length ? { events: context.crashEvents } : undefined,
    warnings
  };
}

function summarizeMediaCore(snapshot: NativeMediaCoreStateSnapshot, state?: ProductionState): SupportBundleMediaCore {
  const recordingManifest = state ? buildRecordingManifest(state, snapshot) : undefined;
  return {
    sceneId: snapshot.sceneId,
    renderPlanId: snapshot.renderPlan.renderPlanId,
    source: {
      adapterId: snapshot.sourceSnapshot.adapterId,
      kind: snapshot.sourceSnapshot.kind,
      status: snapshot.sourceSnapshot.status,
      subscribedSourceCount: snapshot.sourceSnapshot.subscribedSourceCount,
      droppedFrameCount: snapshot.sourceSnapshot.droppedFrameCount,
      lowResolutionFrameCount: snapshot.sourceSnapshot.lowResolutionFrameCount,
      issues: (snapshot.sourceSnapshot.issues ?? []).map((issue) => ({
        sourceId: issue.sourceId,
        participantId: issue.participantId,
        displayName: issue.displayName,
        health: issue.health,
        severity: issue.severity,
        detail: issue.detail
      }))
    },
    compositor: {
      status: snapshot.compositor.status,
      programFrameCount: snapshot.compositor.programFrameCount,
      droppedFrameCount: snapshot.compositor.droppedFrameCount,
      degradedFrameCount: snapshot.compositor.degradedFrameCount
    },
    transport: {
      status: snapshot.programTransport.status,
      frameNumber: snapshot.programTransport.frameNumber,
      latencyMs: snapshot.programTransport.latencyMs
    },
    encoder: {
      status: snapshot.encoderSession.status,
      lifecycle: snapshot.encoderSession.lifecycle.status,
      targetCount: snapshot.encoderSession.targets.length
    },
    senders: {
      status: snapshot.outputSenderSession.status,
      activeSenderCount: snapshot.outputSenderSession.activeSenderCount,
      destinations: snapshot.outputSenderSession.senders.map((sender) => ({
        destination: sender.destination,
        status: sender.status,
        startedAtMs: sender.startedAtMs,
        stoppedAtMs: sender.stoppedAtMs,
        lastFrameNumber: sender.lastFrameNumber,
        framesSent: sender.framesSent,
        retryCount: sender.retryCount,
        latencyMs: sender.latencyMs,
        bitrateMbps: sender.bitrateMbps,
        warning: sender.warning
      }))
    },
    recording: snapshot.recording
      ? {
          status: snapshot.recording.status,
          writerStatus: snapshot.recording.writerStatus,
          totalFramesWritten: snapshot.recording.totalFramesWritten,
          totalDroppedFrames: snapshot.recording.totalDroppedFrames,
          estimatedDiskRateMBps: snapshot.recording.estimatedDiskRateMBps,
          streams: snapshot.recording.streams.map((stream) => ({
            kind: stream.kind,
            participantId: stream.participantId,
            status: stream.status,
            readiness: stream.readiness,
            expectedFrames: stream.expectedFrames,
            framesWritten: stream.framesWritten,
            missingFrames: stream.missingFrames,
            droppedFrames: stream.droppedFrames,
            bytesWritten: stream.bytesWritten,
            warning: stream.warning
          }))
        }
      : undefined,
    recordingManifest: recordingManifest
      ? {
          manifestId: recordingManifest.manifestId,
          sessionId: recordingManifest.sessionId,
          status: recordingManifest.status,
          active: recordingManifest.active,
          sceneId: recordingManifest.sceneId,
          targetFolder: recordingManifest.targetFolder,
          filenamePrefix: recordingManifest.filenamePrefix,
          format: recordingManifest.format,
          quality: recordingManifest.quality,
          outputProfile: recordingManifest.outputProfile,
          trackCount: recordingManifest.tracks.length,
          isoTrackCount: recordingManifest.tracks.filter((track) => track.kind === "iso").length,
          estimatedTotalMbps: recordingManifest.isoPlan.estimatedTotalMbps,
          estimatedFileSizeGbPerHour: recordingManifest.isoPlan.estimatedFileSizeGbPerHour,
          totals: recordingManifest.totals,
          warnings: recordingManifest.warnings
        }
      : undefined,
    operatorActions: snapshot.operatorActions.map((action) => ({
      actionId: action.actionId,
      severity: action.severity,
      area: action.area,
      title: action.title,
      detail: action.detail,
      command: action.command,
      relatedId: action.relatedId
    })),
    eventLog: snapshot.eventLog.map((event) => ({
      eventId: event.eventId,
      atMs: event.atMs,
      severity: event.severity,
      area: event.area,
      title: event.title,
      detail: event.detail,
      relatedId: event.relatedId,
      commandType: event.commandType
    })),
    warnings: snapshot.warnings
  };
}

function getDestinationStates(state: ProductionState): OutputDestinationState[] {
  return state.outputDestinations.map((destination) => {
    const sessionDestination = state.outputSession.destinations.find((item) => item.id === destination.id);
    return sessionDestination ?? { ...destination, active: false, health: "idle", bitrateMbps: 0 };
  });
}

function countDuplicateAssignments(scenes: SceneTemplate[]) {
  return scenes.reduce((count, scene) => {
    const participantSlots = scene.routes
      ? scene.routes
          .filter((route) => (route.mode === "fixed" || route.mode === "spotlight") && route.participantId)
          .map((route) => route.participantId as string)
      : (scene.slots ?? []).filter((slot) => slot !== "screen-share");
    return count + participantSlots.length - new Set(participantSlots).size;
  }, 0);
}

function countUnavailableScreenShare(state: ProductionState) {
  const activeScene = state.scenes.find((scene) => scene.id === state.activeSceneId);
  const sceneNeedsShare =
    activeScene?.layout === "speaker-slides" ||
    activeScene?.slots?.includes("screen-share") ||
    activeScene?.routes?.some((route) => route.mode === "screen-share");
  const hasShare = state.participants.some((participant) => participant.isScreenSharing);
  return sceneNeedsShare && !hasShare ? 1 : 0;
}

function buildWarnings(
  state: ProductionState,
  duplicateAssignments: number,
  unavailableScreenShare: number,
  lowDeliveredResolution: number
) {
  const warnings: string[] = [];

  if (duplicateAssignments > 0) {
    warnings.push(`${duplicateAssignments} duplicate scene assignment${duplicateAssignments === 1 ? "" : "s"} detected.`);
  }

  if (unavailableScreenShare > 0) {
    warnings.push("Program scene expects screen share, but no participant is sharing.");
  }

  if (lowDeliveredResolution > 0) {
    warnings.push(`${lowDeliveredResolution} participant feed${lowDeliveredResolution === 1 ? "" : "s"} delivered below target resolution.`);
  }

  if (state.outputSession.health.network === "warning") {
    warnings.push("Output network health is warning.");
  }

  state.captionOverlay.warnings.forEach((warning) => warnings.push(warning));

  return warnings;
}

function buildTriageLines(
  state: ProductionState,
  warnings: string[],
  isoCapacity: SupportBundle["isoCapacity"],
  context: SupportBundleContext
) {
  const lines = [
    `Show: ${state.meetingTitle} (${state.mode})`,
    `Program: ${state.activeSceneId}; Preview: ${state.previewSceneId}`,
    `Participants: ${state.participants.length}; low-quality feeds: ${state.participants.filter((participant) => participant.health !== "live").length}`,
    `Outputs: ${state.outputSession.statusText}; active destinations: ${state.outputSession.activeDestinationIds.length}`,
    `Output warnings: ${warnings.length}`,
    `ISO runway: ${isoCapacity.recordingRunwayMinutes} minutes at ${isoCapacity.estimatedPathCount} estimated paths.`,
    "SDK subscribe errors: 0"
  ];

  if (context.runtime) {
    lines.push(`Runtime: ${context.runtime.label} (${context.runtime.status})`);
    if (context.runtime.restartCount > 0) {
      lines.push(`Media core restarts: ${context.runtime.restartCount}`);
    }
  }

  if (context.crashEvents?.length) {
    const latest = context.crashEvents[context.crashEvents.length - 1];
    lines.push(`Latest crash: exit ${latest.exitCode ?? "unknown"} at ${latest.at}`);
  }

  return lines;
}

function estimateIsoCapacity(state: ProductionState, freeDiskBytes: number): SupportBundle["isoCapacity"] {
  const selectedParticipantIds = state.recordingSettings.isoParticipantIds.filter((participantId) =>
    state.participants.some((participant) => participant.id === participantId && participant.health !== "video-off")
  );
  const estimatedPathCount = Math.max(1, selectedParticipantIds.length + 1);
  const estimatedDiskRateMBps = Number((estimatedPathCount * 1.85).toFixed(2));
  const recordingRunwayMinutes = Math.floor(freeDiskBytes / (estimatedDiskRateMBps * 1024 * 1024) / 60);
  const warning = recordingRunwayMinutes < 30 ? "Recording runway under 30 minutes." : undefined;

  return {
    selectedParticipantIds,
    estimatedPathCount,
    estimatedDiskRateMBps,
    freeDiskBytes,
    recordingRunwayMinutes,
    warning
  };
}

function getParticipantAction(participant: Participant) {
  const actions: Record<FeedHealth, string> = {
    live: participant.isMuted ? "Check mute state if the participant should be audible." : "No action needed.",
    "low-resolution": "Ask participant to improve network or reduce competing bandwidth.",
    recovering: "Keep routed only if needed; verify the feed stabilizes before taking live.",
    "video-off": "Keep as audio-only or ask participant to enable video."
  };

  return actions[participant.health];
}

function redactEndpoint(endpoint: string) {
  if (!endpoint) {
    return "";
  }

  try {
    const url = new URL(endpoint);
    if (url.username) {
      url.username = "redacted";
    }
    if (url.password) {
      url.password = "redacted";
    }
    ["key", "token", "passphrase", "password", "secret", "code"].forEach((name) => {
      if (url.searchParams.has(name)) {
        url.searchParams.set(name, "redacted");
      }
    });
    return url.toString();
  } catch {
    return endpoint.replace(/(token|key|passphrase|password|secret|code)=([^&\s]+)/gi, "$1=redacted");
  }
}

function redactSecret(value?: string) {
  return value ? "present-redacted" : "absent";
}

function slugify(value: string) {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/(^-|-$)/g, "") || "corevideo";
}

File: native-shell/CoreVideoPro.MediaCore/Models/SupportBundle.cs
namespace CoreVideoPro.MediaCore.Models;

/// <summary>
/// REDACTED diagnostics + crash/recovery snapshot for support escalation.
/// C# port of the React shell's SupportBundle (src/domain/production.ts).
/// No secrets (stream keys, endpoints, URLs with credentials) are emitted.
/// </summary>
public sealed record SupportBundle
{
    public required string Id { get; init; }
    public required string CreatedAt { get; init; }
    public required SupportBundleAppSummary App { get; init; }
    public required string SummaryText { get; init; }
    public IReadOnlyList<string> TriageLines { get; init; } = [];
    public IReadOnlyList<SupportBundleOutputDestinationSummary> Outputs { get; init; } = [];
    public SupportBundleMediaCore? MediaCore { get; init; }
    public SupportBundleRuntime? Runtime { get; init; }
    public SupportBundleCrashRecovery? CrashRecovery { get; init; }
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record SupportBundleAppSummary
{
    public required string Name { get; init; }
    public required string Version { get; init; }
    public required string Platform { get; init; }
}

public sealed record SupportBundleRuntime
{
    public required string Status { get; init; }
    public required string Label { get; init; }
    public required string Host { get; init; }
    public required string Platform { get; init; }
    public int RestartCount { get; init; }
    public bool Recovering { get; init; }
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record SupportBundleCrashRecovery
{
    public IReadOnlyList<SupportBundleCrashEvent> Events { get; init; } = [];
}

public sealed record SupportBundleCrashEvent
{
    public required string At { get; init; }
    public int? ExitCode { get; init; }
    public int RestartCount { get; init; }
}

public sealed record SupportBundleOutputDestinationSummary
{
    public required string Id { get; init; }
    public string Name { get; init; } = string.Empty;
    public string Protocol { get; init; } = "rtmp";
    public bool Enabled { get; init; }
    public bool Active { get; init; }
    public string? Resolution { get; init; }
    public string? Fps { get; init; }
    public string? Codec { get; init; }
    public string? EncoderMode { get; init; }
    public double? TargetBitrateMbps { get; init; }
    public string? Format { get; init; }
    public string? Quality { get; init; }
    /// <summary>Redacted endpoint/URL — credentials and secret query params stripped.</summary>
    public string Endpoint { get; init; } = string.Empty;
    /// <summary>Always "absent" or "present-redacted" — the raw key is never emitted.</summary>
    public string StreamKey { get; init; } = "absent";
}

public sealed record SupportBundleMediaCore
{
    public string? SceneId { get; init; }
    public string? RenderPlanId { get; init; }
    public required SupportBundleMediaCoreSource Source { get; init; }
    public required SupportBundleMediaCoreCompositor Compositor { get; init; }
    public required SupportBundleMediaCoreTransport Transport { get; init; }
    public required SupportBundleMediaCoreEncoder Encoder { get; init; }
    public required SupportBundleMediaCoreAudio Audio { get; init; }
    public required SupportBundleMediaCoreSenders Senders { get; init; }
    public SupportBundleMediaCoreRecording? Recording { get; init; }
    public IReadOnlyList<SupportBundleOperatorAction> OperatorActions { get; init; } = [];
    public IReadOnlyList<SupportBundleEvent> EventLog { get; init; } = [];
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record SupportBundleMediaCoreSource
{
    public required string AdapterId { get; init; }
    public required string Kind { get; init; }
    public required string Status { get; init; }
    public int SubscribedSourceCount { get; init; }
    public int DeliveredFrameCount { get; init; }
    public int LiveFrameCount { get; init; }
    public int StaleFrameCount { get; init; }
    public required string Backing { get; init; }
    public int DroppedFrameCount { get; init; }
    public int LowResolutionFrameCount { get; init; }
}

public sealed record SupportBundleMediaCoreCompositor
{
    public required string Status { get; init; }
    public int ProgramFrameCount { get; init; }
    public int DroppedFrameCount { get; init; }
    public int DegradedFrameCount { get; init; }
}

public sealed record SupportBundleMediaCoreTransport
{
    public required string Status { get; init; }
    public int? FrameNumber { get; init; }
    public double LatencyMs { get; init; }
}

public sealed record SupportBundleMediaCoreEncoder
{
    public required string Status { get; init; }
    public required string Lifecycle { get; init; }
    public int TargetCount { get; init; }
}

public sealed record SupportBundleMediaCoreAudio
{
    public required string Status { get; init; }
    public int MasterLevel { get; init; }
    public double LoudnessLufs { get; init; }
    public bool LimiterActive { get; init; }
    public int MixedFrameCount { get; init; }
    public bool MonitorEnabled { get; init; }
    public string? MonitorStatus { get; init; }
    public string? MonitorDeviceName { get; init; }
    public long MonitorFramesPlayed { get; init; }
    public int ParticipantCount { get; init; }
    public int MutedParticipantCount { get; init; }
    public int BoostedParticipantCount { get; init; }
    public int DuckedParticipantCount { get; init; }
    public int LimitedParticipantCount { get; init; }
    public int LowLevelParticipantCount { get; init; }
    public int PeakOutputLevel { get; init; }
    public int WarningCount { get; init; }
    public IReadOnlyList<string> WarningCategories { get; init; } = [];
    public IReadOnlyList<string> Warnings { get; init; } = [];
    public SupportBundleMediaCoreAudioRouting Routing { get; init; } = new();
    public SupportBundleMediaCoreCaptureAudio Capture { get; init; } = new();
}

public sealed record SupportBundleMediaCoreAudioRouting
{
    public string Status { get; init; } = "idle";
    public int RoutedSendCount { get; init; }
    public int RoutedSourceCount { get; init; }
    public int ProgramTapFrames { get; init; }
    public IReadOnlyList<SupportBundleMediaCoreAudioBusTap> BusTaps { get; init; } = [];
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record SupportBundleMediaCoreAudioBusTap
{
    public required string BusId { get; init; }
    public int Channels { get; init; }
    public int Frames { get; init; }
    public double PeakDbfs { get; init; }
    public double RmsDbfs { get; init; }
}

public sealed record SupportBundleMediaCoreCaptureAudio
{
    public string Status { get; init; } = "idle";
    public int SourceCount { get; init; }
    public int PairedCount { get; init; }
    public int StreamingCount { get; init; }
    public long CaptureFramesReceived { get; init; }
    public int RoutedMasterFrames { get; init; }
    public int RoutedMonitorFrames { get; init; }
    public int FallbackMonitorFrames { get; init; }
    public long MonitorFramesPlayed { get; init; }
    public IReadOnlyList<SupportBundleMediaCoreCaptureAudioSource> Sources { get; init; } = [];
    public IReadOnlyList<string> Warnings { get; init; } = [];
}

public sealed record SupportBundleMediaCoreCaptureAudioSource
{
    public required string CaptureDeviceId { get; init; }
    public string? SourceId { get; init; }
    public string? AudioDeviceName { get; init; }
    public string? AudioSourceKind { get; init; }
    public bool Embedded { get; init; }
    public int AudioSyncOffsetMs { get; init; }
    public bool Paired { get; init; }
    public bool CaptureStreaming { get; init; }
    public long CaptureFramesReceived { get; init; }
    public long EmptyPacketPolls { get; init; }
    public int CaptureSampleRate { get; init; }
    public int CaptureChannels { get; init; }
    public string? EndpointName { get; init; }
    public string? LastError { get; init; }
    public string? Warning { get; init; }
}

public sealed record SupportBundleMediaCoreSenders
{
    public required string Status { get; init; }
    public int ActiveSenderCount { get; init; }
    public IReadOnlyList<SupportBundleMediaCoreSender> Destinations { get; init; } = [];
}

public sealed record SupportBundleMediaCoreSender
{
    public required string Destination { get; init; }
    public required string Status { get; init; }
    public double? StartedAtMs { get; init; }
    public double? StoppedAtMs { get; init; }
    public int? LastFrameNumber { get; init; }
    public int FramesSent { get; init; }
    public int RetryCount { get; init; }
    public double LatencyMs { get; init; }
    public double BitrateMbps { get; init; }
    public string? Warning { get; init; }
}

public sealed record SupportBundleMediaCoreRecording
{
    public required string Status { get; init; }
    public required string WriterStatus { get; init; }
    public int TotalFramesWritten { get; init; }
    public int TotalDroppedFrames { get; init; }
    public double EstimatedDiskRateMBps { get; init; }
    public IReadOnlyList<SupportBundleMediaCoreRecordingStream> Streams { get; init; } = [];
}

public sealed record SupportBundleMediaCoreRecordingStream
{
    public required string Kind { get; init; }
    public string? ParticipantId { get; init; }
    public required string Status { get; init; }
    public int FramesWritten { get; init; }
    public int DroppedFrames { get; init; }
    public long BytesWritten { get; init; }
    public string? Warning { get; init; }
}

public sealed record SupportBundleOperatorAction
{
    public required string ActionId { get; init; }
    public required string Severity { get; init; }
    public required string Area { get; init; }
    public required string Title { get; init; }
    public required string Detail { get; init; }
    public string? Command { get; init; }
    public string? RelatedId { get; init; }
}

public sealed record SupportBundleEvent
{
    public required string EventId { get; init; }
    public double AtMs { get; init; }
    public required string Severity { get; init; }
    public required string Area { get; init; }
    public required string Title { get; init; }
    public required string Detail { get; init; }
    public string? RelatedId { get; init; }
    public string? CommandType { get; init; }
}
